wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolver.java

```diff
@@ -115,15 +116,30 @@ public FactoryDto createFactory(@NotNull final Map<String, String> factoryParame
                 .withPath("/".concat(githubUrl.getRepository()));
           });
     } else if (factory.getDevfile() == null) {
-      // initialize default devfile and github project
+      // initialize default devfile
       factory.setDevfile(urlFactoryBuilder.buildDefaultDevfile(githubUrl.getRepository()));
+    }
+
+    List<ProjectDto> projects = factory.getDevfile().getProjects();
+    // if no projects set, set the default one from GitHub url
+    if (projects.isEmpty()) {
       factory
           .getDevfile()
           .setProjects(
               Collections.singletonList(
                   newDto(ProjectDto.class)
                       .withSource(githubSourceStorageBuilder.buildDevfileSource(githubUrl))
                       .withName(githubUrl.getRepository())));
+    } else {
+      // update existing project with same repository, set current branch if needed
+      projects.forEach(
+          project -> {
+            final String location = project.getSource().getLocation();
+            if (location.equals(githubUrl.repositoryLocation())
```

Comment:
Can `source` or `location` be null? I suppose that devfile with project without location is invalid, but neither `ProjectDto` or `SourceDto` gives us guarantee that it's not null.

Reply:
it is validated after creation so in't supposed to be invalid.